internal/object/objects.go

```diff
@@ -0,0 +1,97 @@
+package object
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icinga-go-library/database"
+	"github.com/icinga/icinga-go-library/types"
+	"github.com/icinga/icinga-notifications/internal/utils"
+	"github.com/pkg/errors"
+	"sync"
+)
+
+var (
+	cache   = make(map[string]*Object)
+	cacheMu sync.Mutex
+)
+
+// DeleteFromCache deletes the Object from the global cache store matching the given ID (if any).
+func DeleteFromCache(id types.Binary) {
+	cacheMu.Lock()
+	defer cacheMu.Unlock()
+
+	delete(cache, id.String())
+}
+
+// RestoreMutedObjects restores all muted objects and their extra / ID tags from the database.
+// Note, this function only retrieves muted objects without non-recovered incident.
+//
+// Returns an error on any database failure and panics when trying to cache an object that's already in the cache store.
+func RestoreMutedObjects(ctx context.Context, db *database.DB) error {
+	clause := `WHERE mute_reason IS NOT NULL AND NOT EXISTS((SELECT 1 FROM incident WHERE object_id = object.id AND recovered_at IS NULL))`
```

Comment:
Isn't this almost the same function as `RestoreObjects()`, just with a different `where` clause?

Also note that the code that you moved from that function into `restoreObjectsTags()` implicitly relies on this following:

https://github.com/Icinga/icinga-notifications/blob/fb1264c826742d4d98ff09f37044a3be81af6412/internal/incident/incidents.go#L63

Your code will likely run into query size limits with many muted objects there.

I've been thinking about this for a while now and didn't come up with a great solution so far. As `RestoreObjects()` uses `utils.ForEachRow()` internally and relies on the external chunking, it's not extended easily. One option might be to just select the IDs here with the `where` clause from above, chunk them and then simply feed them into the existing `RestoreObjects()` function.

Reply:
> One option might be to just select the IDs here with the where clause from above, chunk them and then simply feed them into the existing RestoreObjects() function.

Wouldn't that be a bit tedious to just SELECT the object IDs so that we can use them for a where clause and retrieve the same set of info? If in doubt, I would start two goroutines, similar to how it's done for incidents!

Reply:
As I said, I'm not completely happy with that suggestion, but at least it would allow reusing the existing code instead of duplicating `RestoreObjects()` as `RestoreMutedObjects()` where you than have inlined `utils.ForEachRow()` so that you could adapt the `where` clause and then moved the remaining common code into another function.

Other idea: split `utils.ForEachRow()` in the middle so that it becomes two function, one for generating a `SELECT ... WHERE ... IN (...)` query and another one for executing a query and iterating over the result scanned into instances of a specific type. Other existing users could then get a function that combines the two like it currently does. For the restore object functions, this would allow building a common function like `RestoreObjectsFromQuery()` that could then be used with two different queries.

Reply:
>  For the restore object functions, this would allow building a common function like RestoreObjectsFromQuery() that could then be used with two different queries.

I do not see how this is supposed to solve the specific problem you mentioned above "Your code will likely run into query size limits with many muted objects there." This is just another way of deduplicating existing codes, but does not address the shortcomings of the current state of this PR.

Reply:
> If in doubt, I would start two goroutines, similar to how it's done for incidents!

I've implemented this option!

Reply:
It doesn't. These functions would still have to handle this properly.

There's a second thing here that I think the code structure can be improved: for one, there's some duplication that might not be necessary with more helpful utility functions (hence the suggestion so split `utils.ForEachRow()`). Also the `restoreObjectsTags()` function feels a bit strange to me, like to me it looks like the only reason it exists in the way it does is because you had that common code in two places. Why does it have parameters `objects map[string]*Object` and `ids []types.Binary` where `ids` should contain the keys from `objects`? Probably because that's what existed as local variables. Why does it fetch the tags and then insert the object into the cache? Probably because that was the common code in two functions. I don't think you would come up with that structure when asking questions more like "is this a useful function?" (in the sense that you can describe the purpose of the function not just by "does the steps it does" but also "this is a useful task as a whole").

Reply:
> It doesn't. These functions would still have to handle this properly.

Which functions are you talking about? Also, what do you expect from *handle this correctly*? Should the functions simply fail if they exceed the query limit? If you're suggesting to rely entirely on `utils.ForEachRow` in both cases regardless if it's split or not, how should it *correctly* handle query limits.

Reply:
>  Why does it have parameters `objects map[string]*Object` and `ids []types.Binary` where `ids` should contain the keys from `objects`? Probably because that's what existed as local variables.

Existed as local variables? That's because the objects are only added to the global cache once they have been fully restored and thus are passed on to functions where they are needed.

> Why does it fetch the tags and then insert the object into the cache? Probably because that was the common code in two functions.

No. But this is where the objects get fully restored. I don't know what exactly you want to achieve with this, but for me this is just a simple (internal) helper function that does what it is supposed to do.

Reply:
The current version of your `RestoreMutedObjects()` function can easily be extended into one that can be used to restore objects by ID as well. Have a look at the following code. `restoreObjectsFromQuery()` is basically that that function with `query` and `args` exposed as function arguments. The other two functions then call it with different queries.

For the `RestoreObjectsById()` case - assuming it's never called with more than `MaxPlaceholdersPerStatement` IDs as it's currently the case - the bulker won't do much, it will just generate a single chunk, but having it doesn't hurt and allows using the very same code.

```go
// RestoreMutedObjects restores all muted objects and their extra / ID tags from the database.
// Note, this function only retrieves muted objects without non-recovered incident.
//
// Returns an error on any database failure and panics when trying to cache an object that's already in the cache store.
func RestoreMutedObjects(ctx context.Context, db *database.DB) error {
	query := db.BuildSelectStmt(new(Object), new(Object)) + " WHERE mute_reason IS NOT NULL " +
		"AND NOT EXISTS((SELECT 1 FROM incident WHERE object_id = object.id AND recovered_at IS NULL))"
	return restoreObjectsFromQuery(ctx, db, query)
}

// RestoreObjectsById restores all objects and their (extra)tags matching the given IDs from the database.
// Returns error on any database failures and panics when trying to cache an object that's already in the cache store.
func RestoreObjectsById(ctx context.Context, db *database.DB, ids []types.Binary) error {
	var obj *Object
	query, args, err := sqlx.In(db.BuildSelectStmt(obj, obj)+" WHERE id IN (?)", ids)
	if err != nil {
		return err
	}

	return restoreObjectsFromQuery(ctx, db, query, args...)
}

// restoreObjectsFromQuery takes a query that returns rows of the object table, executes it and loads the returned
// objects into the local cache.
//
// Returns an error on any database failure and panics when trying to cache an object that's already in the cache store.
func restoreObjectsFromQuery(ctx context.Context, db *database.DB, query string, args ...any) error {
	objects := make(chan *Object)
	g, ctx := errgroup.WithContext(ctx)
	g.Go(func() error {
		defer close(objects)

		err := utils.ExecAndApply[Object](ctx, db, query, args, func(o *Object) {
			o.db = db
			o.Tags = map[string]string{}
			o.ExtraTags = map[string]string{}

			select {
			case objects <- o:
			case <-ctx.Done(): // The ctx error is going to be handled by utils.ExecAndApply.
			}
		})

		return errors.Wrap(err, "cannot restore muted objects without an active incident")
	})

	g.Go(func() error {
		bulks := com.Bulk(ctx, objects, db.Options.MaxPlaceholdersPerStatement, com.NeverSplit[*Object])

		for {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case bulk, ok := <-bulks:
				if !ok {
					return nil
				}

				g.Go(func() error {
					var ids []types.Binary
					objectsMap := map[string]*Object{}
					for _, obj := range bulk {
						objectsMap[obj.ID.String()] = obj
						ids = append(ids, obj.ID)
					}

					// Restore object ID tags matching the given object ids
					err := utils.ForEachRow[IdTagRow](ctx, db, "object_id", ids, func(ir *IdTagRow) {
						objectsMap[ir.ObjectId.String()].Tags[ir.Tag] = ir.Value
					})
					if err != nil {
						return errors.Wrap(err, "cannot restore objects ID tags")
					}

					// Restore object extra tags matching the given object ids
					err = utils.ForEachRow[ExtraTagRow](ctx, db, "object_id", ids, func(et *ExtraTagRow) {
						objectsMap[et.ObjectId.String()].ExtraTags[et.Tag] = et.Value
					})
					if err != nil {
						return errors.Wrap(err, "cannot restore objects extra tags")
					}

					addObjectsToCache(objectsMap)

					return nil
				})
			}
		}
	})

	return g.Wait()
}
```